From e25f6234f70ca2b2dfb1a2447826ec255b82c33b Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 9 May 2012 22:55:59 +0300
Subject: [PATCH 159/311] conf/obj.h: describe pinned objects and private
 fields

+ Demote c2_conf_obj::co_type to a private "rank" with intention
  to get rid of this field in the future.
---
 conf/obj.c |  3 ++-
 conf/obj.h | 25 ++++++++++++++++++++++---
 2 files changed, 24 insertions(+), 4 deletions(-)

diff --git a/conf/obj.c b/conf/obj.c
index 419b046..ea6f173 100644
--- a/conf/obj.c
+++ b/conf/obj.c
@@ -207,9 +207,10 @@
  * <!---------------------------------------------------------------->
  * @subsection conf-lspec-state State Specification
  *
- * - @ref conf-fspec-obj-status
  * - @ref confc-lspec-state "States of a confc state machine"
  * - @ref confd-lspec-state "States of a FOM at confd side"
+ * - @ref conf-fspec-obj-status
+ * - @ref conf-fspec-obj-pinned
  *
  * <hr> <!------------------------------------------------------------>
  * @section conf-conformance Conformance
diff --git a/conf/obj.h b/conf/obj.h
index 88dfdb9..35bc410 100644
--- a/conf/obj.h
+++ b/conf/obj.h
@@ -32,6 +32,8 @@
  * - @ref conf-fspec-obj-ovw
  * - @ref conf-fspec-obj-status
  * - @ref conf-fspec-obj-enum
+ * - @ref conf-fspec-obj-pinned
+ * - @ref conf-fspec-obj-private
  *
  * <hr> <!------------------------------------------------------------>
  * @section conf-fspec-obj-ovw Overview
@@ -81,6 +83,21 @@
  * - c2_conf_status --- readiness of the configuration data contained
  *   in an object.
  *
+ * <hr> <!------------------------------------------------------------>
+ * @section conf-fspec-obj-pinned Pinned Objects
+ *
+ * If object's reference counter is non-zero, the object is said to be
+ * @em pinned. Stubs cannot be pinned, only C2_CS_READY objects can.
+ * The reference counter (c2_conf_obj::co_nrefs) is used by confc
+ * library, it is not supposed to be accessed by the application.
+ *
+ * <hr> <!------------------------------------------------------------>
+ * @section conf-fspec-obj-private Private Fields
+ *
+ * Some fields of configuration object data structures are marked as
+ * private. They are used by confc and confd and should not be
+ * accessed by applications.
+ *
  * @see @ref conf_dfspec_obj "Detailed Functional Specification"
  */
 
@@ -127,12 +144,14 @@ enum c2_conf_objtype {
 /**
  * Generic configuration object.
  *
- * This structure is embedded as a first member into each concrete
- * configuration object.
+ * This structure consists of fields that are common to all
+ * configuration objects. It is embedded as a first member into each
+ * concrete configuration object.
  */
 struct c2_conf_obj {
-	enum c2_conf_objtype co_type; /* XXX try to get rid of this field */
 	struct c2_conf_buf   co_key;
+/* private: internal use only */
+	enum c2_conf_objtype co_type; /* XXX try to get rid of this field */
 	enum c2_conf_status  co_status;
 	/**
 	 * Reference counter.
-- 
1.8.3.2

